zebra-rpc/src/sync.rs

```diff
@@ -0,0 +1,385 @@
+                        "failed to commit block to non-finalized state"
+                    );
+
+                    break false;
+                }
+
+                while self
+                    .non_finalized_state
+                    .best_chain_len()
+                    .expect("just successfully inserted a non-finalized block above")
+                    > MAX_BLOCK_REORG_HEIGHT
+                {
+                    tracing::trace!("finalizing block past the reorg limit");
+                    self.non_finalized_state.finalize();
+                }
+
+                self.update_channels(block);
+                current_tip_hash = block_hash;
+                last_chain_tip_hash = current_tip_hash;
+
+                // If the block hash matches the output from the `getbestblockhash` RPC method, we can wait until
+                // the best block hash changes to get the next block.
+                if block_hash == target_tip_hash {
+                    break false;
+                }
+            };
+
+            if should_reset_non_finalized_state {
+                self.try_catch_up_with_primary().await;
+                let block = self.finalized_chain_tip_block().await.expect(
+                    "should have genesis block after successful bestblockheightandhash response",
+                );
+
+                last_chain_tip_hash = block.hash;
+                self.non_finalized_state =
+                    NonFinalizedState::new(&self.non_finalized_state.network);
+                self.update_channels(block);
+            }
+        }
+    }
+
+    /// Tries to catch up to the primary db instance for an up-to-date view of finalized blocks.
+    // TODO: Use getblock RPC if it fails to catch up to primary?
+    async fn try_catch_up_with_primary(&self) {
+        let db = self.db.clone();
+        tokio::task::spawn_blocking(move || {
+            if let Err(catch_up_error) = db.try_catch_up_with_primary() {
+                tracing::warn!(?catch_up_error, "failed to catch up to primary");
+            }
+        })
+        .wait_for_panics()
+        .await
+    }
+
+    /// If the non-finalized state is empty, tries to catch up to the primary db instance for
+    /// an up-to-date view of finalized blocks.
+    ///
+    /// Returns true if the non-finalized state is empty and the target hash is in the finalized state.
+    async fn is_finalized_tip_change(&self, target_tip_hash: block::Hash) -> bool {
+        self.non_finalized_state.chain_count() == 0 && {
+            let db = self.db.clone();
+            tokio::task::spawn_blocking(move || {
+                if let Err(catch_up_error) = db.try_catch_up_with_primary() {
+                    tracing::warn!(?catch_up_error, "failed to catch up to primary");
+                }
+                db.contains_hash(target_tip_hash)
+            })
+            .wait_for_panics()
+            .await
+        }
+    }
+
+    /// Returns the current tip height and hash
```

Comment:
```suggestion
    /// Returns the current tip hash and the next height immediately after the current tip height
```

Reply:
Applied in https://github.com/ZcashFoundation/zebra/pull/8596/commits/15feed173daced582340fdcdb4d4bb8cc9593a39.